### ISSUE TITLE
jenkins/bug: request handle memory leaks in lock_dt etc.

### ISSUE THREAD

**Issue by hzhou** [pmodels/mpich#3863]
During PR #3840, when we were tweaking dtpool algorithm, the following tests fails:
```
rma/lock_contention_dt 4  -type=MPI_INT -count=65530 -seed=149 -testsize=16 
rma/lock_dt 2  -type=MPI_INT -count=65530 -seed=161 -testsize=16 
rma/lock_dt_flush 2  -type=MPI_INT -count=65530 -seed=173 -testsize=16 
rma/lock_dt_flushlocal 2  -type=MPI_INT -count=65530 -seed=185 -testsize=16 
rma/lockall_dt 4  -type=MPI_INT -count=65530 -seed=197 -testsize=16 
rma/lockall_dt_flush 4  -type=MPI_INT -count=65530 -seed=209 -testsize=16 
rma/lockall_dt_flushall 4  -type=MPI_INT -count=65530 -seed=221 -testsize=16 
rma/lockall_dt_flushlocal 4  -type=MPI_INT -count=65530 -seed=233 -testsize=16 
rma/lockall_dt_flushlocalall 4  -type=MPI_INT -count=65530 -seed=245 -testsize=16
```
Example error message:
```
not ok 945 - ./rma/lock_contention_dt 4

## Test output (expected 'No Errors'):
##  No Errors
## [0] 232 at [0x000000000223c0e0], rc/mpid/ch4/netmod/include/../ofi/ofi_am_impl.h[123]
## [0] 232 at [0x0000000002239920], rc/mpid/ch4/netmod/include/../ofi/ofi_am_impl.h[123]
## [0] 168 at [0x0000000002234c80], rc/mpid/ch4/netmod/include/../ofi/ofi_am_impl.h[123]
## [1] 136 at [0x00000000029802f0], rc/mpid/ch4/netmod/include/../ofi/ofi_am_impl.h[123]
## [1] 232 at [0x0000000002976b30], rc/mpid/ch4/netmod/include/../ofi/ofi_am_impl.h[123]
## [2] 168 at [0x0000000002163f10], rc/mpid/ch4/netmod/include/../ofi/ofi_am_impl.h[123]    
```
I only did a cursory look, here is my note:

> With certain datatype (created by certain DTPool count/seed combination), a request handle initialized in src/mpid/ch4/netmod/ofi/ofi_am_impl.h:MPIDI_OFI_am_init_request is never cleared with matching MPIDI_OFI_am_clear_request.

**Comment by hzhou** [comment]
To reproduce: 
* `./configure --with-device=ch4:ofi --enable-g=all --enable-nemesis-dbg-localoddeven`
* export `MPIR_CVAR_ODD_EVEN_CLIQUES=1`
* in `test/mpi/rma`: `mpirun -n 2 -l ./lock_dt -type=MPI_INT -count=4090 -seed=16 -testsize=1`

* `seed=16` generates following datatypes:
```
[0] orig_obj: nesting levels: 0
[0] base: MPI_INT
[0]
[0] target_obj: nesting levels: 1
[0] 0: hvector [numblks 2, blklen 409, stride 1644]
[0] base: MPI_INT
[0]
[0] result_obj: nesting levels: 0
[0] base: MPI_INT
```
* 4090 is the lowest count to trigger error.  Next counts that can trigger error: 4095, 4098, 4102, 4103, 4105, 4109, 4110, 4113, 4115

**Comment by hzhou** [comment]
In `ofi`, we allocate `hdr_ptr` in the request object and need to hold the buffer until `ack` come back. However, in handling `ack`, we simply throw away that buffer without the proper clear procedure. Since we use pool buffer as much as we can, the memory leak only manifests when the needed header buffer is bigger than a limit -- 128 bytes currently.

**Comment by hzhou** [comment]
Fixed in PR #3840
